Stripe/STPRedirectContext.m

```diff
@@ -25,9 +25,59 @@
 
 typedef void (^STPBoolCompletionBlock)(BOOL success);
 
-@interface STPRedirectContext () <SFSafariViewControllerDelegate, STPURLCallbackListener>
+/*
+ SFSafariViewController sometimes manages its own dismissal and does not currently provide
+ any easier API hooks to detect when the dismissal has completed. This machinery exists to
+ insert ourselves into the View Controller transitioning process and detect when a dismissal
+ transition has completed.
+*/
+@protocol STPSafariViewControllerDismissalDelegate <NSObject>
+- (void)safariViewControllerDidCompleteDismissal:(SFSafariViewController *)controller;
+@end
+
+@interface STPSafariViewControllerPresentationController : UIPresentationController
+@property (nonatomic, weak, nullable) id<STPSafariViewControllerDismissalDelegate> dismissalDelegate;
+@end
+
+@implementation STPSafariViewControllerPresentationController
+- (void)dismissalTransitionDidEnd:(BOOL)completed {
+    if ([self.presentedViewController isKindOfClass:[SFSafariViewController class]]) {
+        [self.dismissalDelegate safariViewControllerDidCompleteDismissal:(SFSafariViewController *)self.presentedViewController];
+    }
+    return [super dismissalTransitionDidEnd:completed];
+}
+@end
+
+@interface STPSafariViewControllerTransitioningDelegate : NSObject <UIViewControllerTransitioningDelegate>
```

Comment:
Could we make `STPRedirectContext` conform to `UIViewControllerTransitioningDelegate` and remove this middleman?